File: src/game/systems/waypointmovement.lua
local vector = require 'vendor.vector'
local Aspect = require 'ecs.aspect'
local System = require 'ecs.system'
local Attack = require 'game.components.attack'
local Fixture = require 'game.components.fixture'
local Movement = require 'game.components.movement'
local Position = require 'game.components.position'
local Waypoint = require 'game.components.waypoint'

local aspect = Aspect:new({Fixture, Movement, Position, Waypoint}, {Attack})
local WaypointMovement = System:new('waypointmovement', aspect)

local function getCurrent(waypoint)
  return waypoint.waypoints[waypoint.current]
end

local function getNext(waypoint)
  if waypoint.current == #waypoint.waypoints then
    return waypoint.waypoints[1]
  else
    return waypoint.waypoints[waypoint.current + 1]
  end
end

local function advance(waypoint)
  if waypoint.current == #waypoint.waypoints then
    waypoint.current = 1
  else
    waypoint.current = waypoint.current + 1
  end
end

local function shouldAdvance(axis, velocity, position, next)
  if not next[axis] then
    return true
  elseif
    velocity > 0 and position[axis] >= next[axis] or
      velocity < 0 and position[axis] <= next[axis] or
      -- this is a hacky
      math.abs(position[axis] - next[axis]) < 1
   then
    return true
  else
    return false
  end
end

function WaypointMovement:update(dt)
  for _, entity in pairs(self.entities) do
    local fixture = entity:as(Fixture)
    local pos = entity:as(Position)
    local waypoint = entity:as(Waypoint)
    local body = fixture.fixture:getBody()
    local velocityX, velocityY = body:getLinearVelocity()
    local current = getCurrent(waypoint)
    local next = getNext(waypoint)
    local shouldAdvanceX = shouldAdvance('x', velocityX, pos, next)
    local shouldAdvanceY = shouldAdvance('y', velocityY, pos, next)
    local newVelocityX = 0
    local newVelocityY = 0
    local angle = math.atan2((next.y or pos.y) - pos.y, next.x - pos.x)

    body:setGravityScale(1)

    if shouldAdvanceX and shouldAdvanceY then
      advance(waypoint)
    else
      if not shouldAdvanceX then
        newVelocityX = math.cos(angle) * waypoint.speed
      end

      if next.y then
        body:setGravityScale(0)

        if not shouldAdvanceY then
          newVelocityY = math.sin(angle) * waypoint.speed
        end
      else
        newVelocityY = velocityY
      end
    end

    body:setLinearVelocity(newVelocityX, newVelocityY)
  end
end

return WaypointMovement

File: client2/src/game/systems/WaypointMovement.ts
import { Aspect } from '../../ecs/Aspect'
import { System } from '../../ecs/System'
import { GameObject } from '../components/GameObject'
import { Movement } from '../components/Movement'
import { Position } from '../components/Position'
import { Point, Waypoint } from '../components/Waypoint'
import { SystemFlag } from '../flags'

type Axis = 'x' | 'y'

const shouldAdvance = (
  axis: Axis,
  velocity: number,
  from: Point,
  to: Partial<Point>
) => {
  const coordinate = to[axis]

  if (!coordinate) {
    return true
  }

  if (
    (velocity > 0 && from[axis] >= coordinate) ||
    (velocity < 0 && from[axis] <= coordinate) ||
    Math.abs(from[axis] - coordinate) < 1
  ) {
    return true
  }

  return false
}

const advance = (waypoint: Waypoint) => {
  if (waypoint.current + 1 === waypoint.path.length) {
    waypoint.current = 0
  } else {
    waypoint.current++
  }
}

const getVelocityX = (angle: number, velocity: number) =>
  math.cos(angle) * velocity

const getVelocityY = (angle: number, velocity: number) =>
  math.sin(angle) * velocity

export class WaypointMovementSystem extends System {
  static _id = 'WaypointMovement'
  _id = WaypointMovementSystem._id

  static _flag = SystemFlag.WaypointMovement
  _flag = WaypointMovementSystem._flag

  static _aspect = new Aspect(
    [GameObject, Movement, Position, Waypoint],
    [
      /*Attack*/
    ]
  )
  _aspect = WaypointMovementSystem._aspect

  update = (dt: number) => {
    this.entities.forEach(entity => {
      entity.manager
      const waypoint = entity.as(Waypoint)
      const gameObject = entity.as(GameObject)
      const position = entity.as(Position)

      if (!waypoint || !gameObject || !position) {
        return
      }

      if (!waypoint.active || !waypoint.path.length) {
        return
      }

      const point = waypoint.path[waypoint.current]

      const angle = math.atan2(
        (point.y || position.y) - position.y,
        (point.x || position.x) - position.x
      )

      const body = gameObject.fixture.getBody()

      let [velocityX, velocityY] = body.getLinearVelocity()

      const shouldAdvanceX = shouldAdvance('x', velocityX, position, point)

      const shouldAdvanceY = shouldAdvance('y', velocityY, position, point)

      if (shouldAdvanceX && shouldAdvanceY) {
        advance(waypoint)
      } else {
        if (!shouldAdvanceX) {
          velocityX = getVelocityX(angle, waypoint.speed)
        }

        if (point.y === null) {
          body.setGravityScale(1)
        } else {
          body.setGravityScale(0)

          if (!shouldAdvanceY) {
            velocityY = getVelocityY(angle, waypoint.speed)
          }
        }
      }

      body.setLinearVelocity(velocityX, velocityY)
    })
  }
}
